rfd/0003-plugins-unknown-events.md

```diff
@@ -0,0 +1,384 @@
+---
+authors: Tiago Silva (tiago.silva@goteleport.com)
+state: draft
+---
+
+# RFD 3 - Event handler unknown events
+
+## Required Approvers
+
+* Engineering: @r0mant && ( @rosstimothy || @edoardo )
+
+## What
+
+Customers want to deploy Teleport event handlers that can successfully handle
+all types of events. Currently, Teleport does not provide a way to handle new
+events that are not known to the event handler. This RFD proposes a way to
+handle unknown events when the Teleport event handler version is not up to date
+with the Teleport Auth server version.
+
+## Why
+
+The Teleport Auth server and Teleport event handler are two separate components
+that are deployed independently. This means that the event handler can be
+deployed with a version that is not up to date with the Auth server version.
+
+When the event handler is not up to date with the Auth server, it will not be
+able to handle new events that are not known to the event handler. This can
+cause the event handler to crash (previous behavior) or ignore the unknown events.
+
+This is problematic because the event handler is a critical component that
+should be able to handle all types of events. If the event handler is not able
+to handle new events, it can miss important events and drift from the Auth server truth.
+
+The Teleport event handler receives events from the Auth server, serializes them
+using JSON and sends them to the upstream service (e.g. Splunk, FluentD, etc).
+When the event handler is not up to date with the Auth server, the new events
+will not be known to the event handler and the protobuf raw cannot be deserialized.
+Without the deserialized event, the event handler cannot serialize the event
+using JSON and send it to the upstream service. We could send the raw protobuf
+to auth server and have the auth server serialize it to JSON but this is not
+ideal because it will add a lot of overhead to the Auth server and will add
+a lot of latency to the event handler.
+
+Currently, the event handler ignores unknown events.
+
+## Proposal
+
+The proposal consists of having Auth server serve the events as JSON when the
+client requests the JSON endpoints. This will allow the event handler to receive
+the events as JSON and send them to the upstream service without having to
+deserialize the Protobuf and serialize it again to JSON.
+
+To achieve this, we will add two new endpoints to the Auth server that will return
+the events as JSON. The event handler will use these endpoints to retrieve the events
+and session events as JSON.
+
+The new endpoints will be:
```

Comment:
It doesn't change the nature of this proposal as it's just an implementation detail, but we've been trying to migrate away from the monolithic protobuf service approach and into split services, so maybe this should serve as the start of the `teleport.auditlog.v1.AuditLogService`?